spotbugs/src/main/java/edu/umd/cs/findbugs/BugPattern.java

```diff
@@ -79,9 +82,10 @@
      * @param bugsUrl
      *            URL of web-page containing bug descriptions or null if there's no such page.
      */
-    public BugPattern(String type, String abbrev, String category, boolean experimental, String shortDescription,
+    public BugPattern(String ruleId, String type, String abbrev, String category, boolean experimental, String shortDescription,
             String longDescription, String detailText, String bugsUrl, int cweid) {
 
+        this.ruleId = ruleId;
```

Comment:
If we do not allow plugins to provide no `ruleId` attribute, it's better to check null-ness of given value and throw a `NullPointerException` with a user-friendly message. Also it's better to annotate the parameter with [`@NonNull`](https://spotbugs.readthedocs.io/en/stable/annotations.html#edu-umd-cs-findbugs-annotations-nonnull) (even though other existing parameters aren't annotated so :<).

Reply:
@KengoTODA thank you for pointing out the plugin mechanism, you are right we should keep it backward compatible. 
in case plugin doesn't pass rule id we can use the BugPattern type as rule id, which is currently used as rule id. and provide new API interface so plugin owners can add the opaque rule id in future.

@michaelcfanning @eddynaka what do you think?

Reply:
Thanks.

If we need to follow the [definition of 'stable value'](https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#def_stable_value) and cannot make this attribute optional, it is also possible to release a major version to break compatibility. Of course we need to help plugin authors to migrate, though.

Reply:
Ideally, we would break plug-in authors on update so that the opportunity to provide a valid opaque id is very obvious, but also provide a very easy way to fall back to the legacy rule id if they can't stop to make a fix. @yongyan-gh, please go read the SARIF spec on the `deprecatedIds` property for SARIF `reportingDescriptor` objects. If we are emitting the `rules` table to Spotbugs SARIF, this data can be populated. If we do not emit this data, obviously, we can't persist the deprecated ids. Doing this is not a high priority, as I doubt there are SARIF readers out there who consume this information, though. :) People worried about compatibility will just consult the readable name, which maps to the old rule id value.

Reply:
@KengoTODA in latest change we added an overload constructor to wrap the new constructor and marked with deprecated annotation. 
In this way it wont break existing plugins. Plugins call deprecated constructor will get a warning to use new constructor and provide opaque rule id. 
Please let me know what do you think?